consensus/src/sync/peer_list.rs

```diff
@@ -67,13 +70,17 @@ impl<N: Network> Clone for PeerList<N> {
         Self {
             peers_set: self.peers_set.clone(),
             peers: self.peers.clone(),
+            notify_nonempty: Default::default(),
         }
     }
 }
 
 impl<N: Network> PeerList<N> {
     pub fn add_peer(&mut self, peer_id: N::PeerId) -> bool {
         if self.peers_set.insert(peer_id) {
+            if self.peers.is_empty() {
```

Comment:
> This seems weird, why does the set need to be empty to notify nonempty?

The set only becomes nonempty when it was previously empty.

> Is there a guarantee that you cannot register on this Notify when the peer list is already non empty?

Not directly here, although I guess I could add it. The `notify_nonempty_peers` ensures it by returning an `Option` that only contains a notifier when the set is empty. I guess I should move that guarantee to this function.

> Would there be any harm in just moving this down one line and inverting the condition?
> 
> That way whenever the set gets added to it will trigger a notify_waiters, which there should not be any if it was previously not empty, but if there are they now know it is (still) not empty. Otherwise they will be waiting forever as long as the list never gets empty again.

The harm would mostly be hiding bugs and unnecessary wakeups. We could also wake this from many other places, so it's not clear that "adding a peer to an otherwise nonempty list" should trigger the "peer list became nonempty" any more than, say, "removing a peer from a list, and it's still nonempty".

Reply:
> Is there a guarantee that you cannot register on this Notify when the peer list is already non empty?

I added that guarantee to the API now. Thanks for bringing it up.